keps/672-execution-policy/README.md

```diff
@@ -0,0 +1,398 @@
+                    claimName: model-initializer
+    - name: launcher
+      template:
+        spec:
+          parallelism: 1
+          completions: 1
+          template:
+            spec:
+              containers:
+                - name: launcher
+                  image: docker.io/kubeflow/mpi-launcher
+                  resources:
+                    limits:
+                      nvidia.com/gpu: 5
+                  volumeMounts:
+                    - mountPath: /workspace/dataset
+                      name: dataset-initializer
+                    - mountPath: /workspace/pre-trained-model
+                      name: model-initializer
+              volumes:
+                - name: dataset-initializer
+                  persistentVolumeClaim:
+                    claimName: dataset-initializer
+                - name: model-initializer
+                  persistentVolumeClaim:
+                    claimName: model-initializer
+    - name: trainer-node
+      template:
+        spec:
+          parallelism: 3
+          completions: 3
+          template:
+            spec:
+              containers:
+                - name: trainer
+                  image: docker.io/kubeflow/deepspeed-trainer
+                  resources:
+                    limits:
+                      nvidia.com/gpu: 5
+                  volumeMounts:
+                    - mountPath: /workspace/dataset
+                      name: dataset-initializer
+                    - mountPath: /workspace/pre-trained-model
+                      name: model-initializer
+              volumes:
+                - name: dataset-initializer
+                  persistentVolumeClaim:
+                    claimName: dataset-initializer
+                - name: model-initializer
+                  persistentVolumeClaim:
+                    claimName: model-initializer
+```
+
+#### Story 3
+
+As an HPC user, I want to run a set of simulations with MPI that complete successfully before
+using the data for a next step analysis.
+
+As an HPC user I want to scale up my simulation job only after a small set have completed
+successfully. For that case, I could imagine essentially the same replicatedJob done twice,
+just with a larger size. If the outcome of the first isn't success you wouldn't launch
+the larger bulk of work.
+
+### Risks and Mitigations
+
+This API will not allow to describe DAGs to avoid workflow manager features in JobSet.
+The goal is to only focus on Job sequence to cover model training/HPC use-cases.
+
+## Design Details
+
+### API Details
+
+```golang
+type JobSetSpec struct {
+	ExecutionPolicy *ExecutionPolicy `json:"executionPolicy,omitempty"`
+}
+
+type ExecutionPolicyOption string
+
+const (
+	// This is the default settings.
+	// AnyOrder means that Jobs will be started in any order.
+	AnyOrder ExecutionPolicyOption = "AnyOrder"
+
+	// InOrder starts the ReplicatedJobs in order that they are listed. Jobs within a ReplicatedJob
```

Comment:
curious if this depends on the data structure used in the controller? Is it easy to establish order without a serial number in the spec? 

Could the order be changed while persisting to etcd?

Reply:
I think, this is how currently StartupPolicy API works. If I remember correctly, the InitContainers and Containers also executed in the order how the listed in the YAML. @tenzen-y @kannon92 do you have any thoughts on this ?

Reply:
> curious if this depends on the data structure used in the controller? Is it easy to establish order without a serial number in the spec?
> 
> Could the order be changed while persisting to etcd?

[ReplicatedJobs is represented as a slice data structure](https://github.com/kubernetes-sigs/jobset/blob/b92cbdcbe77f168a82cf12c9f349bdd053d2ae5c/api/jobset/v1alpha2/jobset_types.go#L82C1-L85C66) so the order is preserved, even through serialization, persistence to etcd, and deserialization. 

So operating on replicatedJobs in the order they are listed is straightforward and consistent.